File: backend/component/urlDownloadService.js

const fs = require('fs').promises;
const fsSync = require('fs');
const path = require('path');
const https = require('https');
const http = require('http');

class UrlDownloadService {
  constructor() {
    this.uploadDir = path.join(__dirname, '..', 'uploads');
    this.ensureUploadDir();
  }

  async ensureUploadDir() {
    try {
      await fs.mkdir(this.uploadDir, { recursive: true });
    } catch (error) {
      console.error('❌ Failed to create upload directory:', error);
    }
  }

  /**
   * Download PDF from URL and save to local file system
   * @param {string} url - PDF URL to download
   * @param {string} fileId - Unique file identifier
   * @returns {Promise<Object>} Download result with file path and metadata
   */
  async downloadPDF(url, fileId) {
    return new Promise((resolve, reject) => {
      console.log(`📥 Starting PDF download from: ${url}`);
      
      // Validate URL
      if (!this.isValidUrl(url)) {
        return reject(new Error('Invalid URL format'));
      }

      // Check if it's likely a PDF URL
      if (!this.isPdfUrl(url)) {
        console.warn('⚠️ URL may not be a PDF, attempting download anyway');
      }

      const fileName = `${fileId}_${this.extractFilenameFromUrl(url) || 'document'}.pdf`;
      const filePath = path.join(this.uploadDir, fileName);

      // Choose http or https based on URL
      const client = url.startsWith('https:') ? https : http;

      const request = client.get(url, (response) => {
        // Handle redirects
        if (response.statusCode >= 300 && response.statusCode < 400 && response.headers.location) {
          console.log(`↩️ Following redirect to: ${response.headers.location}`);
          return this.downloadPDF(response.headers.location, fileId)
            .then(resolve)
            .catch(reject);
        }

        if (response.statusCode !== 200) {
          return reject(new Error(`HTTP ${response.statusCode}: ${response.statusMessage}`));
        }

        // Validate content type if available
        const contentType = response.headers['content-type'];
        if (contentType && !contentType.includes('application/pdf') && !contentType.includes('octet-stream')) {
          console.warn(`⚠️ Content-Type is ${contentType}, expected PDF`);
        }

        const fileStream = fsSync.createWriteStream(filePath);
        let downloadedBytes = 0;
        const totalBytes = parseInt(response.headers['content-length']) || 0;

        response.on('data', (chunk) => {
          downloadedBytes += chunk.length;
          if (totalBytes > 0) {
            const progress = ((downloadedBytes / totalBytes) * 100).toFixed(1);
            console.log(`📥 Download progress: ${progress}% (${downloadedBytes}/${totalBytes} bytes)`);
          }
        });

        response.pipe(fileStream);

        fileStream.on('finish', () => {
          fileStream.close();
          console.log(`✅ PDF downloaded successfully: ${fileName} (${downloadedBytes} bytes)`);
          
          // Validate downloaded file
          this.validatePdfFile(filePath)
            .then((isValid) => {
              if (!isValid) {
                console.warn('⚠️ Downloaded file may not be a valid PDF');
              }
              
              resolve({
                success: true,
                filePath: filePath,
                fileName: fileName,
                originalUrl: url,
                size: downloadedBytes,
                mimetype: 'application/pdf'
              });
            })
            .catch((validationError) => {
              console.warn('⚠️ PDF validation failed:', validationError.message);
              // Still resolve as we have the file, just warn about validation
              resolve({
                success: true,
                filePath: filePath,
                fileName: fileName,
                originalUrl: url,
                size: downloadedBytes,
                mimetype: 'application/pdf',
                validationWarning: validationError.message
              });
            });
        });

        fileStream.on('error', (error) => {
          console.error('❌ File write error:', error);
          this.cleanupFile(filePath);
          reject(new Error(`File write failed: ${error.message}`));
        });
      });

      request.on('error', (error) => {
        console.error('❌ Download request error:', error);
        this.cleanupFile(filePath);
        reject(new Error(`Download failed: ${error.message}`));
      });

      request.setTimeout(30000, () => {
        request.destroy();
        this.cleanupFile(filePath);
        reject(new Error('Download timeout (30 seconds)'));
      });
    });
  }

  /**
   * Validate if file is a valid PDF
   * @param {string} filePath - Path to file to validate
   * @returns {Promise<boolean>} Whether file is valid PDF
   */
  async validatePdfFile(filePath) {
    try {
      const buffer = await fs.readFile(filePath);
      
      // Check PDF magic number (first 4 bytes should be %PDF)
      const header = buffer.slice(0, 4).toString();
      if (header !== '%PDF') {
        throw new Error('File does not have PDF header');
      }

      // Check minimum file size (PDFs are typically at least 100 bytes)
      if (buffer.length < 100) {
        throw new Error('File too small to be a valid PDF');
      }

      return true;
    } catch (error) {
      throw new Error(`PDF validation failed: ${error.message}`);
    }
  }

  /**
   * Extract filename from URL
   * @param {string} url - URL to extract filename from
   * @returns {string} Extracted filename without extension
   */
  extractFilenameFromUrl(url) {
    try {
      const urlObj = new URL(url);
      const pathname = urlObj.pathname;
      const filename = path.basename(pathname);
      
      if (filename && filename !== '/') {
        // Remove extension for our naming
        return path.parse(filename).name;
      }
      
      // Fallback: use domain name
      return urlObj.hostname.replace(/\./g, '_');
    } catch (error) {
      return 'unknown_document';
    }
  }

  /**
   * Check if URL is valid format
   * @param {string} url - URL to validate
   * @returns {boolean} Whether URL is valid
   */
  isValidUrl(url) {
    try {
      new URL(url);
      return url.startsWith('http://') || url.startsWith('https://');
    } catch (error) {
      return false;
    }
  }

  /**
   * Check if URL likely points to a PDF
   * @param {string} url - URL to check
   * @returns {boolean} Whether URL likely points to PDF
   */
  isPdfUrl(url) {
    const lowerUrl = url.toLowerCase();
    return lowerUrl.includes('.pdf') || 
           lowerUrl.includes('pdf') || 
           lowerUrl.includes('application/pdf');
  }

  /**
   * Clean up file if download fails
   * @param {string} filePath - Path to file to clean up
   */
  async cleanupFile(filePath) {
    try {
      if (fsSync.existsSync(filePath)) {
        await fs.unlink(filePath);
        console.log(`🧹 Cleaned up failed download: ${filePath}`);
      }
    } catch (error) {
      console.warn('⚠️ Failed to cleanup file:', filePath, error.message);
    }
  }

  /**
   * Get download statistics
   * @returns {Object} Download statistics
   */
  getStats() {
    return {
      uploadDir: this.uploadDir,
      serviceName: 'UrlDownloadService',
      supportedProtocols: ['http', 'https'],
      maxTimeout: 30000
    };
  }
}

module.exports = new UrlDownloadService();

File: backend_python/component/url_download_service.py

import os
import asyncio
import aiofiles
import aiohttp
from pathlib import Path
from urllib.parse import urlparse
import re


class URLDownloadService:
    def __init__(self):
        self.upload_dir = Path(__file__).parent.parent / 'uploads'
        self.ensure_upload_dir()

    def ensure_upload_dir(self):
        """Ensure upload directory exists"""
        try:
            self.upload_dir.mkdir(parents=True, exist_ok=True)
        except Exception as error:
            print(f'❌ Failed to create upload directory: {error}')

    async def check_pdf_size(self, url: str, max_size_mb: int = 10) -> dict:
        """
        Check PDF size via HEAD request before downloading
        @param url: PDF URL to check
        @param max_size_mb: Maximum allowed size in MB
        @returns: Size check result with metadata
        """
        print(f'📏 Checking PDF size for: {url}')
        
        max_size_bytes = max_size_mb * 1024 * 1024  # Convert MB to bytes
        
        try:
            async with aiohttp.ClientSession(timeout=aiohttp.ClientTimeout(total=10)) as session:
                async with session.head(url, allow_redirects=True) as response:
                    if response.status != 200:
                        # If HEAD fails, try GET with range to get size
                        async with session.get(url, headers={'Range': 'bytes=0-0'}, allow_redirects=True) as range_response:
                            if range_response.status in [206, 200]:
                                content_range = range_response.headers.get('content-range', '')
                                if content_range and '/' in content_range:
                                    try:
                                        total_size = int(content_range.split('/')[-1])
                                    except ValueError:
                                        # Fallback: get content-length from headers
                                        total_size = int(range_response.headers.get('content-length', 0))
                                else:
                                    total_size = int(range_response.headers.get('content-length', 0))
                            else:
                                raise Exception(f'Cannot determine file size. HTTP {range_response.status}')
                    else:
                        # Use content-length from HEAD response
                        total_size = int(response.headers.get('content-length', 0))
                    
                    # Check content type if available
                    content_type = response.headers.get('content-type', '')
                    
                    size_mb = total_size / (1024 * 1024)
                    
                    print(f'📏 PDF size check result: {size_mb:.2f}MB (limit: {max_size_mb}MB)')
                    
                    return {
                        'success': True,
                        'size_bytes': total_size,
                        'size_mb': round(size_mb, 2),
                        'within_limit': total_size <= max_size_bytes,
                        'content_type': content_type,
                        'max_size_mb': max_size_mb
                    }
                    
        except Exception as error:
            print(f'❌ Size check failed: {error}')
            return {
                'success': False,
                'error': str(error),
                'size_bytes': 0,
                'size_mb': 0,
                'within_limit': False,
                'content_type': '',
                'max_size_mb': max_size_mb
            }

    async def download_pdf(self, url: str, file_id: str) -> dict:
        """
        Download PDF from URL and save to local file system
        @param url: PDF URL to download
        @param file_id: Unique file identifier
        @returns: Download result with file path and metadata
        """
        print(f'📥 Starting PDF download from: {url}')

        # Validate URL
        if not self.is_valid_url(url):
            raise ValueError('Invalid URL format')

        # Check if it's likely a PDF URL
        if not self.is_pdf_url(url):
            print('⚠️ URL may not be a PDF, attempting download anyway')

        file_name = f"{file_id}_{self.extract_filename_from_url(url) or 'document'}.pdf"
        file_path = self.upload_dir / file_name

        try:
            async with aiohttp.ClientSession(timeout=aiohttp.ClientTimeout(total=30)) as session:
                async with session.get(url, allow_redirects=True) as response:
                    if response.status != 200:
                        raise Exception(f'HTTP {response.status}: {response.reason}')

                    # Validate content type if available
                    content_type = response.headers.get('content-type', '')
                    if content_type and 'application/pdf' not in content_type and 'octet-stream' not in content_type:
                        print(f'⚠️ Content-Type is {content_type}, expected PDF')

                    downloaded_bytes = 0
                    total_bytes = int(response.headers.get('content-length', 0))

                    async with aiofiles.open(file_path, 'wb') as file:
                        async for chunk in response.content.iter_chunked(8192):
                            await file.write(chunk)
                            downloaded_bytes += len(chunk)
                            
                            if total_bytes > 0:
                                progress = (downloaded_bytes / total_bytes) * 100
                                print(f'📥 Download progress: {progress:.1f}% ({downloaded_bytes}/{total_bytes} bytes)')

            print(f'✅ PDF downloaded successfully: {file_name} ({downloaded_bytes} bytes)')

            # Validate downloaded file
            try:
                is_valid = await self.validate_pdf_file(file_path)
                if not is_valid:
                    print('⚠️ Downloaded file may not be a valid PDF')
            except Exception as validation_error:
                print(f'⚠️ PDF validation failed: {validation_error}')

            return {
                'success': True,
                'filePath': str(file_path),
                'fileName': file_name,
                'originalUrl': url,
                'size': downloaded_bytes,
                'mimetype': 'application/pdf'
            }

        except Exception as error:
            print(f'❌ Download failed: {error}')
            await self.cleanup_file(file_path)
            raise Exception(f'Download failed: {str(error)}')

    async def validate_pdf_file(self, file_path: Path) -> bool:
        """
        Validate if file is a valid PDF
        @param file_path: Path to file to validate
        @returns: Whether file is valid PDF
        """
        try:
            async with aiofiles.open(file_path, 'rb') as file:
                header = await file.read(4)

            # Check PDF magic number (first 4 bytes should be %PDF)
            if header.decode('utf-8', errors='ignore') != '%PDF':
                raise Exception('File does not have PDF header')

            # Check minimum file size (PDFs are typically at least 100 bytes)
            if file_path.stat().st_size < 100:
                raise Exception('File too small to be a valid PDF')

            return True
        except Exception as error:
            raise Exception(f'PDF validation failed: {str(error)}')

    def extract_filename_from_url(self, url: str) -> str:
        """
        Extract filename from URL
        @param url: URL to extract filename from
        @returns: Extracted filename without extension
        """
        try:
            parsed_url = urlparse(url)
            pathname = parsed_url.path
            filename = os.path.basename(pathname)

            if filename and filename != '/':
                # Remove extension for our naming
                return os.path.splitext(filename)[0]

            # Fallback: use domain name
            return parsed_url.hostname.replace('.', '_') if parsed_url.hostname else 'unknown_document'
        except Exception:
            return 'unknown_document'

    def is_valid_url(self, url: str) -> bool:
        """
        Check if URL is valid format
        @param url: URL to validate
        @returns: Whether URL is valid
        """
        try:
            result = urlparse(url)
            return all([result.scheme, result.netloc]) and result.scheme in ['http', 'https']
        except Exception:
            return False

    def is_pdf_url(self, url: str) -> bool:
        """
        Check if URL likely points to a PDF
        @param url: URL to check
        @returns: Whether URL likely points to PDF
        """
        lower_url = url.lower()
        return '.pdf' in lower_url or 'pdf' in lower_url or 'application/pdf' in lower_url

    async def cleanup_file(self, file_path: Path):
        """
        Clean up file if download fails
        @param file_path: Path to file to clean up
        """
        try:
            if file_path.exists():
                file_path.unlink()
                print(f'🧹 Cleaned up failed download: {file_path}')
        except Exception as error:
            print(f'⚠️ Failed to cleanup file: {file_path}, {error}')

    def get_stats(self) -> dict:
        """
        Get download statistics
        @returns: Download statistics
        """
        return {
            'uploadDir': str(self.upload_dir),
            'serviceName': 'URLDownloadService',
            'supportedProtocols': ['http', 'https'],
            'maxTimeout': 30000
        }
